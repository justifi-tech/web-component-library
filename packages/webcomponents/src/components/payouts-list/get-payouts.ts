import { Payout } from '../../api';
import { getErrorMessage } from '../../api/services/utils';

export const makeGetPayouts =
  ({ id, authToken, service }) =>
  async ({ params, onSuccess, onError }) => {
    try {
      const response = await service.fetchPayouts(id, authToken, params);

      if (!response.error) {
        const pagingInfo = {
          ...response.page_info,
        };

        const payouts =
          response.data?.map((dataItem) => new Payout(dataItem)) || [];

        onSuccess({ payouts, pagingInfo });
      } else {
        const responseError = getErrorMessage(response.error);
        return onError(responseError);
      }
    } catch (error) {
      return onError(error.message || error);
    }
  };
