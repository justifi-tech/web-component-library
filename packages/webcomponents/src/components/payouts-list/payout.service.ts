import { Api, IApiResponseCollection, IPayout } from '../../api';
import { config } from '../../../config';

export interface IPayoutService {
  fetchPayouts(
    accountId: string,
    authToken: string,
    params: any
  ): Promise<IApiResponseCollection<IPayout[]>>;
}

export class PayoutService implements IPayoutService {
  async fetchPayouts(
    accountId: string,
    authToken: string,
    params: any
  ): Promise<IApiResponseCollection<IPayout[]>> {
    const api = Api(authToken, config.proxyApiOrigin);
    const endpoint = `account/${accountId}/payouts`;
    return api.get(endpoint, params);
  }
}
