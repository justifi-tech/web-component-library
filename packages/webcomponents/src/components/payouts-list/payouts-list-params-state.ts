import { createStore } from '@stencil/store';
import { PayoutsQueryParams } from '../../api';

const filterParamsInitialState = {};
const propsParamsInitialState = {};

const filterParamsStore = createStore<PayoutsQueryParams>(() => filterParamsInitialState);
const { state: filterParams, on: onQueryParamsChange } = filterParamsStore;

const propsParamsStore = createStore<PayoutsQueryParams>(() => propsParamsInitialState);
const { state: propsParams } = propsParamsStore;

const getRequestParams = (): PayoutsQueryParams => {
  return {
    ...filterParams,
    ...propsParams
  };
};

const clearParams = () => {
  filterParamsStore.reset();
}

export {
  filterParams,
  propsParams,
  onQueryParamsChange,
  clearParams,
  getRequestParams
};
