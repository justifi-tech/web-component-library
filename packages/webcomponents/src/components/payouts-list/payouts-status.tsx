
export const MapPayoutStatusToBadge = (status: string) => {
  switch (status) {
    case 'canceled':
      return "<span class='badge bg-danger' title='Transfer to your bank account failed'>Canceled</span>";
    case 'failed':
      return "<span class='badge bg-danger' title='Transfer to your bank account failed'>Failed</span>";
    case 'forwarded':
      return "<span class='badge bg-secondary' title='This payout initially failed; the funds have been forwarded to your next successful payout'>Forwarded</span>";
    case 'in_transit':
      return "<span class='badge bg-primary' title='Transfer to your bank account has been initiated'>In Transit</span>";
    case 'paid':
      return "<span class='badge bg-success' title='Successfully deposited into your bank account'>Paid</span>";
    case 'pending':
      return "<span class='badge bg-primary' title='Batched and scheduled to be transferred'>Pending</span>";
    case 'scheduled':
      return "<span class='badge bg-primary' title='Batched and scheduled to be transferred'>Scheduled</span>";
    case 'withdrawn':
      return "<span class='badge bg-success' title='Negative payout balance successfully withdrawn from your bank account'>Withdrawn</span>";
  };
};
