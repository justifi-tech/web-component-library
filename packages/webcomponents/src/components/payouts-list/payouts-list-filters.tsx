import { Component, h, Prop } from '@stencil/core';
import { filterParams, propsParams, clearParams } from './payouts-list-params-state';
import { StyledHost } from '../../ui-components';
import { convertToLocal, convertToUTC } from '../../utils/utils';
import { createdAfterPayoutsListFilterParam, payoutsListFilterMenu } from '../../styles/parts';
@Component({
  tag: 'justifi-payouts-list-filters',
  shadow: true
})
export class PayoutsListFilters {
  @Prop() createdAfter?: string;
  @Prop() createdBefore?: string;

  componentWillLoad() {
    const propsToSet = {
      created_after: this.createdAfter,
      created_before: this.createdBefore
    };

    Object.entries(propsToSet).forEach(([key, value]) => {
      if (value) {
        propsParams[key] = value;
      }
    });
  }
  
  setParamsOnChange = (name: string, value: string) => {
    filterParams[name] = value;
  }

  handleDateInput = (name: string, value: string) => {
    const utcDate = convertToUTC(value, { setEndOfDay: true });
    this.setParamsOnChange(name, utcDate);
  }

  render() {
    const filterMenuParams = { ...filterParams }

    return (
      <StyledHost>
        <table-filters-menu params={filterMenuParams} clearParams={clearParams} part={payoutsListFilterMenu}>
          <div class="grid-cols-2 gap-3 p-1">
            <div class="p-2">
              <form-control-date
                name="created_after"
                label="Created After"
                inputHandler={this.handleDateInput}
                defaultValue={
                  convertToLocal(this.createdAfter, { showInputDate: true }) ||
                  convertToLocal(filterParams.created_after, { showInputDate: true })
                }
                disabled={!!this.createdAfter}
                part={createdAfterPayoutsListFilterParam}
              />
            </div>
            <div class="p-2">
              <form-control-date
                name="created_before"
                label="Created Before"
                inputHandler={this.handleDateInput}
                defaultValue={
                  convertToLocal(this.createdBefore, { showInputDate: true }) ||
                  convertToLocal(filterParams.created_before, { showInputDate: true })
                }
                disabled={!!this.createdBefore}
                part={createdAfterPayoutsListFilterParam}
              />
            </div>
          </div>
        </table-filters-menu>
      </StyledHost>
    );
  }
}
