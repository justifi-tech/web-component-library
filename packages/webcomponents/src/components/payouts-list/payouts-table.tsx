import { h } from '@stencil/core';
import { convertToLocal } from '../../utils/utils';
import { MapPayoutStatusToBadge } from './payouts-status';
import { getAlternateTableCellPart, tableHeadCell } from '../../styles/parts';
import { Payout } from '../../api/Payout';

export const defaultColumnsKeys = 'created_at,amount,status,payments_total,refunds_total,fees_total,other_total,csv';

export const payoutTableColumns = {
  created_at: () => (
    <th part={tableHeadCell} scope="col" title="The date the payout was deposited">
      Created
    </th>
  ),
  sub_account_name: () => (
    <th part={tableHeadCell} scope="col" title="The sub account associated with the payout">
      Sub Account
    </th>
  ),
  payments_total: () => (
    <th part={tableHeadCell} scope="col" title="The total payments amount associated with the payout">
      Payments
    </th>
  ),
  refunds_total: () => (
    <th part={tableHeadCell} scope="col" title="The total refunds amount associated with the payout">
      Refunds
    </th>
  ),
  fees_total: () => (
    <th part={tableHeadCell} scope="col" title="The total fees amount associated with the payout">
      Fees
    </th>
  ),
  other_total: () => (
    <th part={tableHeadCell} scope="col" title="The total other amount">
      Other
    </th>
  ),
  amount: () => (
    <th part={tableHeadCell} scope="col" title="The total amount of the payout">
      Amount
    </th>
  ),
  status: () => (
    <th part={tableHeadCell} scope="col" title="The real-time status of each payout">
      Status
    </th>
  ),
  csv: () => (
    <th part={tableHeadCell} scope="col" title="Export CSV" />
  ),
};

export const payoutTableCells = (downloadCSV) => ({
  created_at: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>
      <div class="fw-bold">{convertToLocal(payout.created_at, { showDisplayDate: true })}</div>
    </td>
  ),
  sub_account_name: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.sub_account_name}</td>
  ),
  payments_total: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.formatPaymentAmount(payout.payments_total, true)}</td>
  ),
  refunds_total: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.formatPaymentAmount(payout.refunds_total, true)}</td>
  ),
  fees_total: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.formatPaymentAmount(payout.fees_total, true)}</td>
  ),
  other_total: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.formatPaymentAmount(payout.other_total, true)}</td>
  ),
  amount: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>{payout.formatPaymentAmount(payout.amount, true)}</td>
  ),
  status: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)} innerHTML={MapPayoutStatusToBadge(payout.status)}></td>
  ),
  csv: (payout: Payout, index: number) => (
    <td part={getAlternateTableCellPart(index)}>
      <a href="#" onClick={(event) => { event.preventDefault(); downloadCSV(payout); }}>
        CSV
      </a>
    </td>
  ),
});
