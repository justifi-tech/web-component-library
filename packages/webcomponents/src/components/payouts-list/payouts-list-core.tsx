import { Component, h, Prop, State, Watch, Event, EventEmitter } from '@stencil/core';
import { PagingInfo, Payout, SubAccount, pagingDefaults } from '../../api';
import { ComponentError, ComponentErrorCodes, ComponentErrorSeverity } from '../../api/ComponentError';
import { TableEmptyState, TableErrorState, TableLoadingState } from '../../ui-components';
import { payoutTableCells, payoutTableColumns } from './payouts-table';
import { Table } from '../../utils/table';
import { queryParams, onQueryParamsChange } from './payouts-list-params-state';

@Component({
  tag: 'payouts-list-core',
})

export class PayoutsListCore {
  @Prop() getPayouts: Function;
  @Prop() getSubAccounts: Function;
  @Prop() getPayoutCSV: Function;
  @Prop() columns: string;

  @State() payouts: Payout[] = [];
  @State() payoutsTable: Table;
  @State() subAccounts: SubAccount[] = [];
  @State() loading: boolean = true;
  @State() errorMessage: string;
  @State() paging: PagingInfo = pagingDefaults;
  @State() pagingParams: any = {};

  @Watch('pagingParams')
  @Watch('getPayouts')
  @Watch('getSubAccounts')
  updateOnPropChange() {
    this.fetchData();
  }

  @Event({
    eventName: 'row-clicked',
    bubbles: true,
  }) rowClicked: EventEmitter<Payout>;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  componentWillLoad() {
    this.payoutsTable = new Table(this.payouts, this.columns, payoutTableColumns, payoutTableCells(this.downloadCSV));
    if (this.getPayouts && this.getSubAccounts) {
      this.fetchData();
    }
 
    onQueryParamsChange('set', () => {
      delete this.pagingParams.before_cursor;
      delete this.pagingParams.after_cursor;
      this.fetchData();
    });

    onQueryParamsChange('reset', () => {
      this.pagingParams = {};
    });
  }

  fetchData(): void {
    this.loading = true;

    this.getPayouts({
      params: this.requestParams,
      onSuccess: ({ payouts, pagingInfo }) => {
        this.payouts = payouts;
        this.paging = pagingInfo;
        this.payoutsTable.collectionData = this.payouts;
        const shouldFetchSubAccounts = this.payoutsTable.columnKeys.includes('sub_account_name');

        if (shouldFetchSubAccounts) {
          this.fetchSubAccounts();
        } else {
          this.loading = false;
        }

      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  async fetchSubAccounts(): Promise<void> {
    this.getSubAccounts({
      params: this.subAccountParams,
      onSuccess: ({ subAccounts }) => {
        this.subAccounts = subAccounts;
        this.payouts = this.payouts.map((payout) => {
          payout.sub_account_name = this.subAccounts.find((subAccount) => subAccount.id === payout.account_id)?.name;
          return payout;
        });
        this.loading = false;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  downloadCSV = (payoutId: string) => {
    this.getPayoutCSV({
      payoutId,
      onError: () => {
        this.errorEvent.emit({
          errorCode: ComponentErrorCodes.FETCH_ERROR,
          message: 'Failed to download CSV',
          severity: ComponentErrorSeverity.ERROR,
        });
      },
    });
  }

  handleClickPrevious = (beforeCursor: string) => {
    this.pagingParams = { before_cursor: beforeCursor };
  };

  handleClickNext = (afterCursor: string) => {
    this.pagingParams = { after_cursor: afterCursor };
  };

  rowClickHandler = (e) => {
    const clickedPayoutID = e.target.closest('tr').dataset.rowEntityId;
    if (!clickedPayoutID) return;
    this.rowClicked.emit(this.payouts.find((payout) => payout.id === clickedPayoutID));
  }

  get requestParams() {
    const combinedParams = { ...queryParams, ...this.pagingParams };
    return combinedParams;
  }

  get subAccountParams() {
    let accountIdNumbers = this.payouts.map((payout) => payout.account_id);
    let uniqueAccountIds = [...new Set(accountIdNumbers)];
    let accountIdString = uniqueAccountIds.join(',');
    return { sub_account_id: accountIdString };
  }

  get entityId() {
    return this.payouts.map((payout) => payout.id);
  }

  get showEmptyState() {
    return !this.loading && !this.errorMessage && this.payoutsTable.rowData.length < 1;
  }

  get showErrorState() {
    return !this.loading && !!this.errorMessage;
  }

  get showRowData() {
    return !this.showEmptyState && !this.showErrorState && !this.loading;
  }

  render() {
    return (
      <div>
        <div class="table-wrapper" part="table-wrapper">
          <table class="table table-hover" part="table">
            <thead class="table-head sticky-top" part="table-head">
              <tr class="table-light text-nowrap" part="table-head-row">
                {this.payoutsTable.columnData.map((column) => column)}
              </tr>
            </thead>
            <tbody class="table-body" part="table-body">
              <TableLoadingState
                  columnSpan={this.payoutsTable.columnKeys.length}
                  isLoading={this.loading}
                />
                <TableEmptyState
                  isEmpty={this.showEmptyState}
                  columnSpan={this.payoutsTable.columnKeys.length}
                />
                <TableErrorState
                  columnSpan={this.payoutsTable.columnKeys.length}
                  errorMessage={this.errorMessage}
                />
                {this.showRowData && this.payoutsTable.rowData.map((data, index) => (
                  <tr
                    data-test-id="table-row"
                    data-row-entity-id={this.entityId[index]}
                    onClick={this.rowClickHandler}
                    part={`table-row ${index % 2 ? "table-row-even" : "table-row-odd"}`}
                  >
                    {data}
                  </tr>
                ))}
            </tbody>
            {this.paging && (
              <tfoot class="sticky-bottom">
                <tr class="table-light align-middle">
                  <td part="pagination-bar" colSpan={this.payoutsTable.columnData.length}>
                    <pagination-menu
                      paging={{
                        ...this.paging,
                        handleClickPrevious: this.handleClickPrevious,
                        handleClickNext: this.handleClickNext,
                      }}
                    />
                  </td>
                </tr>
              </tfoot>
            )}
          </table>
        </div>
      </div>
    );
  }
}
