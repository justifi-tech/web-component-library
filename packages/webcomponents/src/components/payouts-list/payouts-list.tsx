import { Component, h, Prop, State, Watch, Event, EventEmitter } from '@stencil/core';
import { ComponentError, ComponentErrorCodes, ComponentErrorSeverity } from '../../api/ComponentError';
import JustifiAnalytics from '../../api/Analytics';
import { checkPkgVersion } from '../../utils/check-pkg-version';
import { config } from '../../../config';
import { PayoutService } from '../../api/services/payout.service';
import { makeGetPayouts } from './get-payouts';
import { makeGetPayoutCSV } from '../payout-details/get-payout-csv';
import { makeGetSubAccounts } from '../../api/get-subaccounts';
import { SubAccountService } from '../../api/services/subaccounts.service';
import { StyledHost, tableExportedParts } from '../../ui-components';
import { defaultColumnsKeys } from './payouts-table';

/**
  * @exportedPart label: Label for inputs
  * @exportedPart input: The input fields
  * @exportedPart input-invalid: Invalid state for inputs
  * @exportedPart table-head: Table head
  * @exportedPart table-head-row: Head row
  * @exportedPart table-head-cell: Individual head cell
  * @exportedPart table-body: Body of the table
  * @exportedPart table-row: Row of the table
  * @exportedPart table-cell: Individual cell of the table
  * @exportedPart loading-state-cell: Row for loading state
  * @exportedPart loading-state-spinner: Spinner element for loading state
  * @exportedPart error-state: Row for Error state
  * @exportedPart empty-state: Row for Emtpy state
  * @exportedPart pagination-bar: Pagination bar
  * @exportedPart arrow: Both paging buttons
  * @exportedPart arrow-left: Previous page button
  * @exportedPart arrow-right: Next page button
  * @exportedPart button-disabled: Disabled state for paging buttons
  * @exportedPart previous-button-text: Text for Previous button
  * @exportedPart next-button-text: Text for Next button
*/
@Component({
  tag: 'justifi-payouts-list',
  shadow: true,
})

export class PayoutsList {
  @State() getPayouts: Function;
  @State() getPayoutCSV: Function;
  @State() getSubAccounts: Function;
  @State() errorMessage: string = null;
  
  @Prop() accountId: string;
  @Prop() authToken: string;
  @Prop() apiOrigin?: string = config.proxyApiOrigin;
  @Prop() columns?: string = defaultColumnsKeys;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  analytics: JustifiAnalytics;

  componentWillLoad() {
    checkPkgVersion();
    this.analytics = new JustifiAnalytics(this);
    this.initializeGetData();
  }

  disconnectedCallback() {
    this.analytics.cleanup();
  }

  @Watch('accountId')
  @Watch('authToken')
  propChanged() {
    this.initializeGetData();
  }

  private initializeGetData() {
    this.initializePayoutsServices();
    this.initializeGetSubAccounts();
  }

  private initializePayoutsServices() {
    if (this.accountId && this.authToken) {
      const serviceParams = {
        id: this.accountId,
        authToken: this.authToken,
        service: new PayoutService(),
      };
      this.getPayouts = makeGetPayouts(serviceParams);
      this.getPayoutCSV = makeGetPayoutCSV(serviceParams);
    } else {
      this.errorMessage = 'Account ID and Auth Token are required';
      this.errorEvent.emit({
        errorCode: ComponentErrorCodes.MISSING_PROPS,
        message: this.errorMessage,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  }

  private initializeGetSubAccounts() {
    if (this.accountId && this.authToken) {
      this.getSubAccounts = makeGetSubAccounts({
        accountId: this.accountId,
        authToken: this.authToken,
        service: new SubAccountService(),
        apiOrigin: this.apiOrigin
      });
    }
  }

  handleOnError = (event) => {
    this.errorMessage = event.detail.message;
    this.errorEvent.emit(event.detail);
  }

  render() {
    return (
      <StyledHost exportedparts={tableExportedParts}>
        <payouts-list-core
          getPayouts={this.getPayouts}
          getPayoutCSV={this.getPayoutCSV}
          getSubAccounts={this.getSubAccounts}
          onError-event={this.handleOnError}
          columns={this.columns}
        />
      </StyledHost>
    );
  }
}
