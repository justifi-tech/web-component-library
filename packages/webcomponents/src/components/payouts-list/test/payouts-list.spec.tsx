import { newSpecPage } from "@stencil/core/testing";
import { PayoutsList } from "../payouts-list";
import { PayoutsListCore } from "../payouts-list-core";

describe('payouts-list', () => {
  it('renders an error message when accountId and authToken are not provided', async () => {
    const page = await newSpecPage({
      components: [PayoutsList, PayoutsListCore],
      html: '<justifi-payouts-list></justifi-payouts-list>',
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('renders an error message when accountId is not provided', async () => {
    const page = await newSpecPage({
      components: [PayoutsList, PayoutsListCore],
      html: '<justifi-payouts-list auth-token="abc"></justifi-payouts-list>',
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('renders an error message when authToken is not provided', async () => {
    const page = await newSpecPage({
      components: [PayoutsList, PayoutsListCore],
      html: '<justifi-payouts-list account-id="abc"></justifi-payouts-list>',
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });
});
