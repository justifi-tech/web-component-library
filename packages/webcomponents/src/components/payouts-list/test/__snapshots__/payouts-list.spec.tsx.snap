// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payouts-list renders an error message when accountId and authToken are not provided 1`] = `
<justifi-payouts-list>
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
      <div>
        Account ID and Auth Token are required
      </div>
    </main>
  </template>
</justifi-payouts-list>
`;

exports[`payouts-list renders an error message when accountId is not provided 1`] = `
<justifi-payouts-list auth-token="abc">
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
      <div>
        Account ID and Auth Token are required
      </div>
    </main>
  </template>
</justifi-payouts-list>
`;

exports[`payouts-list renders an error message when authToken is not provided 1`] = `
<justifi-payouts-list account-id="abc">
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
      <div>
        Account ID and Auth Token are required
      </div>
    </main>
  </template>
</justifi-payouts-list>
`;
